### PR TITLE
[Import-linter] Find more import errors in import linter pipeline.

### PR DESCRIPTION
**Pull request by crazygao**
# Description

Import linter cannot find items other than static imports.
We can add some more import statement in poetry and find circular import dynamically.

Automatic imports:
Since we have different processes, just dynamic import the module in different process is OK.

image

# All Promptflow Contribution checklist:
- [ ] **The pull request does not introduce [breaking changes].**
- [ ] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [ ] **I have read the [contribution guidelines](../CONTRIBUTING.md).**
- [ ] **Create an issue and link to the pull request to get dedicated review from promptflow team. Learn more: [suggested workflow](../CONTRIBUTING.md#suggested-workflow).**

## General Guidelines and Best Practices
- [ ] Title of the pull request is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### Testing Guidelines
- [ ] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: .github/workflows/promptflow-import-linter.yml [hunk 1]
```diff
@@ -59,4 +59,20 @@ jobs:
...
+    - name: import lint testing private imports from global
+      run: |
+        set -xe
+        rm ${{ env.WORKING_DIRECTORY }}/src/promptflow-tracing/promptflow/__init__.py
+        rm ${{ env.WORKING_DIRECTORY }}/src/promptflow-core/promptflow/__init__.py
+        rm ${{ env.WORKING_DIRECTORY }}/src/promptflow-devkit/promptflow/__init__.py
+        rm ${{ env.WORKING_DIRECTORY }}/src/promptflow-azure/promptflow/__init__.py
+        rm ${{ env.WORKING_DIRECTORY }}/src/promptflow-evals/promptflow/__init__.py
+        cd ${{ env.WORKING_DIRECTORY }}/src/promptflow-azure
+        echo "=== Add more import linter when facing more import errors ==="
```
Comment (zhengfeiwang) [inline .github/workflows/promptflow-import-linter.yml:71]: if we want import-linter detect, we need to manually add here? is it possible to apply an auto-collection for all global import from our repo?

File: .github/workflows/promptflow-import-linter.yml [hunk 2]
```diff
@@ -59,4 +59,20 @@ jobs:
...
+
+        echo "=== promptflow-devkit ==="
+        poetry run python -c $'from promptflow._sdk import *\nfrom promptflow._cli import *'
```
Comment (wangchao1230) [inline .github/workflows/promptflow-import-linter.yml:74]: any strack trace on issue we detected?

Comment (crazygao) [reply .github/workflows/promptflow-import-linter.yml:74]: Yes, if there is error, there will be stack traces here.

For example.
image

File: .github/workflows/promptflow-import-linter.yml (hunk 1 above)
Comment (crazygao) [reply .github/workflows/promptflow-import-linter.yml:71]: Yes and no.
I just suggest that we only test the ones that could be easily broken.
Some imports are embedded inside code, we might need some real statement , eg: PFClient(), to get the real import error.

Comment (crazygao) [reply .github/workflows/promptflow-import-linter.yml:66]: Fixed. 
image

File: scripts/import_linter/import_linter.py [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+            push_file.insert(0, head_tail[1])
+        file = head_tail[0]
+        head_tail = os.path.split(file)
+    push_file.insert(0, "promptflow")
+    return ".".join(push_file)
+
+
+def subprocess_check_python_import(file):
+    print(f'Checking import of {file} on process ID: {os.getpid()}')
+    importlib.import_module(file)
```
Comment (zhengfeiwang) [inline scripts/import_linter/import_linter.py:37]: may add a comment to demonstrate what does `importlib.import_module` consume? and what will happen if it finds some improper import?

File: scripts/import_linter/import_linter.py [hunk 4]
```diff
@@ -0,0 +1,53 @@
+import os
+import subprocess
+import multiprocessing
+import importlib
+
+git_base = subprocess.check_output(['git', 'rev-parse', '--show-toplevel']).decode().strip()
+
+
+def walk_and_ignore_pycache(directory):
```
Comment (zhengfeiwang) [inline scripts/import_linter/import_linter.py:9]: will this walk those swagger generated code? e.g., `src/promptflow-azure/promptflow/azure/_restclient/*`?

**Review by zhengfeiwang — APPROVED** [review]

File: scripts/import_linter/import_linter.py (hunk 4 above)
Comment (crazygao) [reply scripts/import_linter/import_linter.py:9]: Yes

File: scripts/import_linter/import_linter.py (hunk 3 above)
Comment (crazygao) [reply scripts/import_linter/import_linter.py:37]: OK

**Review by zhengfeiwang — DISMISSED** [review]

**Review by 0mza987 — DISMISSED** [review]

**Review by 0mza987 — APPROVED** [review]

**Review by elliotzh — APPROVED** [review]

**Pull request merged**